app/assets/javascripts/webauthn_credentials.js

```diff
@@ -0,0 +1,187 @@
+    const str = atob(base64String)
+    const buffer = new ArrayBuffer(str.length)
+    const byteView = new Uint8Array(buffer)
+    for (let i = 0; i < str.length; i++) {
+      byteView[i] = str.charCodeAt(i)
+    }
+    return buffer
+  }
+
+  const bufferToBase64url = (buffer) => {
+    const byteView = new Uint8Array(buffer)
+    let str = ""
+    for (const charCode of byteView) {
+      str += String.fromCharCode(charCode)
+    }
+    const base64String = btoa(str)
+    const base64urlString = base64String
+      .replace(/\+/g, "-")
+      .replace(/\//g, "_")
+      .replace(/=/g, "")
+    return base64urlString
+  }
+
+  $(FORM_SELECTOR).submit(async (event) => {
+    try {
+      event.preventDefault()
+
+      const form = event.target
+      const nickname = document.querySelector(NICKNAME_INPUT_SELECTOR).value
+
+      let createResponse = await fetch(`${form.action}.json`, {
+        method: "POST",
+        credentials: "same-origin",
+        headers: { "X-CSRF-Token": csrfToken },
+      })
+
+      createResponse = await createResponse.json()
+      createResponse.user.id = base64urlToBuffer(createResponse.user.id)
+      createResponse.challenge = base64urlToBuffer(createResponse.challenge)
+      createResponse.excludeCredentials = createResponse.excludeCredentials.map(
+        (excludeCredential) => {
+          return {
+            id: base64urlToBuffer(excludeCredential.id),
+            type: excludeCredential.type,
+          }
+        }
+      )
+
+      const credentials = await navigator.credentials.create({
+        publicKey: createResponse,
+      })
+
+      let callbackResponse = await fetch(`${form.action}/callback.json`, {
+        method: "POST",
+        credentials: "same-origin",
+        headers: {
+          "X-CSRF-Token": csrfToken,
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          credentials: {
+            type: credentials.type,
+            id: credentials.id,
+            rawId: bufferToBase64url(credentials.rawId),
+            clientExtensionResults: credentials.clientExtensionResults,
+            response: {
+              attestationObject: bufferToBase64url(
+                credentials.response.attestationObject
+              ),
+              clientDataJSON: bufferToBase64url(
+                credentials.response.clientDataJSON
+              ),
+            },
+          },
+          webauthn_credential: { nickname: nickname },
+        }),
+      })
+
+      if (callbackResponse.status == 200) {
+        callbackResponse = await callbackResponse.json()
+        window.location.href = callbackResponse.location
+      } else {
+        callbackResponse = await callbackResponse.json()
+        alert(callbackResponse.message)
+      }
+    } catch (e) {
+      document.querySelector(SUBMIT_SELECTOR).disabled = false
+      throw e
+    }
+  })
+
+  $(SESSION_FORM_SELECTOR).submit(async (e) => {
```

Comment:
maybe I should split this in another file? not sure how I would share the functions for both cases